source/common/tracing/http_tracer_impl.cc

```diff
@@ -113,46 +114,56 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
-  msg->headers().addViaCopy(Http::Headers::get().ContentType, "application/json");
-  msg->headers().addViaCopy(Http::Headers::get().Host, "collector.lightstep.com");
-  msg->headers().addViaCopy("LightStep-Access-Token", access_token);
+void LightStepRecorder::RecordSpan(lightstep::collector::Span&& span) {
+  builder_.addSpan(std::move(span));
 
-  return msg;
+  const int min_span_number_to_send = 5;
+  if (builder_.pendingSpans() == min_span_number_to_send) {
+    lightstep::collector::ReportRequest request;
+    std::swap(request, builder_.pending());
+
+    Grpc::RpcAsyncClientImpl client(sink_->clusterManager());
+    client.send(sink_->collectorCluster(), "lightstep.collector.CollectorService", "Report",
+                std::move(request), *this);
+  }
 }
 
-std::string LightStepUtility::buildJoiningIds(const Http::HeaderMap& request_headers) {
-  std::string join_ids;
+bool LightStepRecorder::FlushWithTimeout(lightstep::Duration) {
+  // Note: We don't expect this to be called, since the Tracer
+  // reference is private to its LightStepSink.
+  return true;
+}
 
-  // Always populate x-request-id as joining id.
-  static const std::string x_request_id_format = R"EOF(
-      {{
-        "TraceKey": "x-request-id",
-        "Value": "{}"
-      }})EOF";
-  join_ids += fmt::format(x_request_id_format, request_headers.get(Http::Headers::get().RequestId));
+std::unique_ptr<lightstep::Recorder>
+LightStepRecorder::NewInstance(LightStepSink* sink, const lightstep::TracerImpl& tracer) {
+  return std::unique_ptr<lightstep::Recorder>(new LightStepRecorder(sink, tracer));
+}
 
-  // Optionally populate x-client-trace-id if present.
-  if (request_headers.has(Http::Headers::get().ClientTraceId)) {
-    static const std::string x_client_trace_id_format = R"EOF(
-      ,{{
-        "TraceKey": "x-client-trace-id",
-        "Value": "{}"
-      }})EOF";
-    join_ids += fmt::format(x_client_trace_id_format,
-                            request_headers.get(Http::Headers::get().ClientTraceId));
+LightStepSink::LightStepSink(const Json::Object& config, Upstream::ClusterManager& cluster_manager,
+                             const std::string& stat_prefix, Stats::Store& stats,
+                             const std::string& service_node,
+                             const lightstep::TracerOptions& options)
+    : collector_cluster_(config.getString("collector_cluster")), cm_(cluster_manager),
+      stats_{LIGHTSTEP_STATS(POOL_COUNTER_PREFIX(stats, stat_prefix + "tracing.lightstep."))},
+      service_node_(service_node), options_(options) {
+  if (!cm_.get(collector_cluster_)) {
+    throw EnvoyException(fmt::format("{} collector cluster is not defined on cluster manager level",
+                                     collector_cluster_));
   }
+}
 
-  return join_ids;
+lightstep::Tracer& LightStepSink::thread_local_tracer() {
+  static thread_local lightstep::Tracer tracer(lightstep::NewUserDefinedTransportLightStepTracer(
```

Comment:
need to use envoy tls here for testing (probably obvious but pointing out anyway)